### PR TITLE
add md5 checksum of the file while exporting highlights

### PR DESCRIPTION
**Pull request by sp4ke**
- This allows interoperability between different readers who want to import highlights on different devices/file systems. All that is needed then to import highlights is to match the local books and their path based on the checksum.

- I needed this to import highlights on sioyek book reader from koreader, I use koreader on tablets and sioyek on my laptop.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9610)

### CODE REVIEW

**Comment by poire-z** [comment]
Pinging @pazos @uroybd for review/merge: is that ok / the right way to solve the use case?

**Comment by pazos** [comment]
I'm not sure about the impact it could have when exporting the entire library with hundred or thousands of books.

I would prefer to add it as an option (default disabled). Everything's is already in place to do so, you'll only need to override the exporter `getMenuTable`.

Have a look at https://github.com/koreader/koreader/blob/83f8c6ed818cfdccc53a41fe85f7a947947bbe34/plugins/exporter.koplugin/target/markdown.lua#L87-L103

**Comment by sp4ke** [comment]
@pazos thanks for the example. I updated the PR accordingly using a menu with the option disabled by default.

File: plugins/exporter.koplugin/target/json.lua [hunk 1]
```diff
@@ -1,13 +1,47 @@
...
 
 -- json exporter
 local JsonExporter = require("base"):new {
     name = "json",
     shareable = Device:canShareText(),
+    init_callback = function(self, settings)
+        local changed = false
+        if settings.bookChecksum == nil then
+            settings.bookChecksum = false
+        end
```
[the remark this answers is not in the document]
Comment (NiLuJe) [reply plugins/exporter.koplugin/target/json.lua:14]: And, perhaps less intuitively, `not nil` yields `true`, so the menu callback also works ;).

File: plugins/exporter.koplugin/target/json.lua [hunk 2]
```diff
@@ -16,6 +50,10 @@ local function format(booknotes)
         file = booknotes.file,
         number_of_pages = booknotes.number_of_pages
     }
+    if settings.bookChecksum then
+        t.file_md5sum = md5.sumFile(booknotes.file)
+
```
Comment (NiLuJe) [inline plugins/exporter.koplugin/target/json.lua:55]: nit: stray LF

File: plugins/exporter.koplugin/target/json.lua [hunk 3]
```diff
@@ -1,13 +1,47 @@
...
+        text = _("Json"),
+        checked_func = function() return self:isEnabled() end,
+        sub_item_table = {
+            {
+                text = _("Export to Json"),
+                checked_func = function() return self:isEnabled() end,
+                callback = function() self:toggleEnabled() end,
+            },
+            {
+                text = _("Show book checksum in export"),
```
Comment (poire-z) [inline plugins/exporter.koplugin/target/json.lua:31]: "Include" rather than "Show" ?

File: plugins/exporter.koplugin/target/json.lua [hunk 4]
```diff
@@ -16,6 +50,10 @@ local function format(booknotes)
         file = booknotes.file,
         number_of_pages = booknotes.number_of_pages
     }
+    if settings.bookChecksum then
```
Comment (pazos) [inline plugins/exporter.koplugin/target/json.lua:53]: something like:

```lua
    local t = {
        title = booknotes.title,
        author = booknotes.author,
        entries = {},
        exported = booknotes.exported,
        file = booknotes.file,
        number_of_pages = booknotes.number_of_pages,
        md5sum = settings.bookChecksum and md5.sumFile(booknotes.file),
    }
```

will get the same output

Comment (sp4ke) [reply plugins/exporter.koplugin/target/json.lua:53]: this exports a field `md5sum = false` is that fine ?

File: plugins/exporter.koplugin/target/json.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (sp4ke) [reply plugins/exporter.koplugin/target/json.lua]: Got it thanks, got used at over checking the conditions in JS

File: plugins/exporter.koplugin/target/json.lua (hunk 4 above)
Comment (pazos) [reply plugins/exporter.koplugin/target/json.lua]: mmm, I guess it happens only when you flip the setting a couple of times.

In any other context I would recommend to use https://github.com/koreader/koreader/blob/c9ff0071e384d3ab7491fc4c4d6da967f4d643de/frontend/luasettings.lua#L156

but here is a bit abstracted away and not easiest thing to do.

You can do something like this on the callback:
```lua
callback = function()
    self.settings.bookChecksum = self.settings.bookChecksum and nil or true
    self:saveSettings()
end,
```

for this PR and we can refactor that bit of code later if some other exporters start to abuse this kind of construct

Comment (sp4ke) [reply plugins/exporter.koplugin/target/json.lua]: I think you misunderstood my last message, I meant the code you proposed worked perfectly as you said, when it's set to `true` the checksum is exported. But when it's set to `false` there is a useless checksum field in the json is set to false. 

The change however on your last comment does not work it makes the option sticky.

Comment (pazos) [reply plugins/exporter.koplugin/target/json.lua]: lol, sorry. sir Boole would be very angry with me :)

Please do what @poire-z suggested, and sorry for the extra time lost :p

Comment (sp4ke) [reply plugins/exporter.koplugin/target/json.lua]: @pazos  @poire-z perfect thanks ! Now the field set to false means `nil` and is actually not exported in json. So this option is incognito until the user needs it :)

Comment (sp4ke) [reply plugins/exporter.koplugin/target/json.lua]: > Please do what @poire-z suggested, and sorry for the extra time lost :p

no pasa nada 😄

**Comment by sp4ke** [comment]
Just made a last push a cleaner comment on the commit, so it can be referred to in the docs/wiki if needed

File: plugins/exporter.koplugin/target/json.lua [hunk 5]
```diff
@@ -1,19 +1,46 @@
 local Device = require("device")
 local rapidjson = require("rapidjson")
+local _ = require("gettext")
+local md5 = require("ffi/MD5")
```
Comment (pazos) [inline plugins/exporter.koplugin/target/json.lua:4]: nit: `md5` before `_`

File: plugins/exporter.koplugin/target/json.lua [hunk 6]
```diff
@@ -1,19 +1,46 @@
...
 
 -- json exporter
 local JsonExporter = require("base"):new {
     name = "json",
     shareable = Device:canShareText(),
 }
 
-local function format(booknotes)
+function JsonExporter:getMenuTable()
+    local menu = {
```
Comment (pazos) [inline plugins/exporter.koplugin/target/json.lua:13]: nit: no need to store the table in a variable if the next thing we're going to do is return it.

you can return a table directly

**Review by pazos — APPROVED** [review]
LGTM

**Pull request merged**